lib/TPP/GPU/LinalgToGpu.cpp

```diff
@@ -30,108 +30,301 @@ using namespace mlir::tpp;
 
 namespace {
 
+// Return true if the operation can be represented with WMMA compute.
+bool supportsMMACompute(linalg::LinalgOp linalgOp) {
+  if (!(isa_and_nonnull<linalg::MatmulOp>(linalgOp) ||
+        isa_and_nonnull<linalg::BatchReduceMatmulOp>(linalgOp)))
+    return false;
+
+  auto aType =
+      linalgOp.getDpsInputOperands()[0]->get().getType().cast<ShapedType>();
+  auto bType =
+      linalgOp.getDpsInputOperands()[1]->get().getType().cast<ShapedType>();
+  auto cType =
+      linalgOp.getDpsInitOperands()[0]->get().getType().cast<ShapedType>();
+
+  ArrayRef<int64_t> shapeA = aType.getShape();
+  ArrayRef<int64_t> shapeC = cType.getShape();
+  int64_t m = shapeC[0];
+  int64_t n = shapeC[1];
+  // Matrix A might be 2D (gemm) or 3D (brgemm) but the last dimension will
+  // always be reduction.
+  int64_t k = shapeA.back();
+
+  // For now, only M-N-K F16[16] x F16[16] x F16[16] WMMA variant is supported.
+  // TODO: add more WMMA combinations.
+  return aType.getElementType().isF16() && bType.getElementType().isF16() &&
+         cType.getElementType().isF16() && m == 16 && n == 16 && k == 16;
+}
+
+// Create WMMA instructions out of matmul-like operation.
+void gemmToGpuMMA(linalg::LinalgOp linalgOp, PatternRewriter &rewriter) {
+  assert((isa_and_nonnull<linalg::MatmulOp>(linalgOp) ||
+          isa_and_nonnull<linalg::BatchReduceMatmulOp>(linalgOp)) &&
+         "Requires a matmul like op for mma lowering");
+
+  Location loc = linalgOp.getLoc();
+
+  auto matA = linalgOp.getDpsInputOperands()[0]->get();
+  auto matB = linalgOp.getDpsInputOperands()[1]->get();
+  auto matC = linalgOp.getDpsInitOperands()[0]->get();
+
+  auto typeA = matA.getType().cast<ShapedType>();
+  auto typeB = matB.getType().cast<ShapedType>();
+  auto typeC = matC.getType().cast<ShapedType>();
+
+  gpu::MMAMatrixType mmaTypeA = gpu::MMAMatrixType::get(
+      typeA.getShape().take_back(2), typeA.getElementType(), "AOp");
+  gpu::MMAMatrixType mmaTypeB = gpu::MMAMatrixType::get(
+      typeB.getShape().take_back(2), typeB.getElementType(), "BOp");
+  gpu::MMAMatrixType mmaTypeC =
+      gpu::MMAMatrixType::get(typeC.getShape(), typeC.getElementType(), "COp");
+
+  // Matrix A might be 2D (gemm) or 3D (brgemm) but the last dimension will
+  // always be reduction.
+  auto ldb = rewriter.getIndexAttr(typeA.getShape().back());
+  auto lda = rewriter.getIndexAttr(typeC.getShape()[0]);
```

Comment:
Are you at tensor level? You need to enforce identity layout otherwise. Looking at shape for the leading dimension is not correct when you have strided memref.
`memref<64x32xf32, strides[512, 1]...>` the leading dim here is 512 not 32.

Reply:
Good point, this can be handled.
Added proper strides support.